### PR TITLE
Windows Support

### PR DESCRIPTION
**Pull request by xorz57**
# Pull Request Status: Ready
# Previous Pull Request: #31 

# Important Changes
- Improved `CMakeLists.txt`.
- Improved `.travis.yml`.
- Added `appveyor.yml`.
- Separated the tests.
- Separated the examples.
- Colored Ouput works on Windows.
- Added support for `std::array`
- Added support for `std::list`

# Windows Build
- MSVC 19.16.27030.1 (Visual Studio 15 2017)
- MSVC 19.21.27702.2 (Visual Studio 16 2019)

# Linux Build
- GCC 7.0
- GCC 8.0
- Clang 8.0.1
- Clang 7.1.0

# macOS Build
- AppleClang 10.0.0.10001145
- AppleClang 10.0.1.10010046

### CODE REVIEW

**Comment by xorz57** [comment]
@sharkdp I finally found my way resolving merge conflicts! Sorry for the mess.

File: include/dbg_macro/dbg.h [hunk 1]
```diff
@@ -269,7 +269,7 @@ pretty_print(std::ostream& stream, Container const& value) {
   const size_t size = value.size();
   const size_t n = std::min(size_t{5}, size);
   size_t i = 0;
-  for (auto it = begin(value); it != end(value) && i < n; ++it, ++i) {
+  for (auto it = std::begin(value); it != std::end(value) && i < n; ++it, ++i) {
```
Comment (hmenke) [inline include/dbg_macro/dbg.h]: This breaks ADL.

Comment (hmenke) [reply include/dbg_macro/dbg.h]: You could put `using std::begin;` and `using std::end;` in front.

Comment (xorz57) [reply include/dbg_macro/dbg.h]: Well, I had no idea. I simply dont use the `using namespace` thing for my reasons.

Comment (xorz57) [reply include/dbg_macro/dbg.h]: I just read [this](https://stackoverflow.com/questions/31604984/is-using-stdbegin-a-good-practice). It doesn't really convince me but I will do it anyway!

Comment (xorz57) [reply include/dbg_macro/dbg.h]: Done

**Comment by xorz57** [comment]
I hope we can merge this one before merge conflicts start appearing again :disappointed:

**Comment by xorz57** [comment]
@sharkdp Since you don't have a `development branch` it would be nice if you pushed a `github release` for each stable version. That way we can use a `direct link` to the header without having to `clone` or `copy paste` the header's source code. Just like what the guys on [Catch2](https://github.com/catchorg/Catch2) did. They have a link on README.md that points to a `github release`, you click it and it starts downloading the header and only that.

**Comment by hmenke** [comment]
> it would be nice if you pushed a `github release` for each stable version

This project is 4 days old.

**Comment by xorz57** [comment]
> > it would be nice if you pushed a `github release` for each stable version
> 
> This project is 4 days old.

I had no idea.

**Comment by xorz57** [comment]
@sharkdp It seems that we have a bug.

```sh
      Start  1: test_array
 1/12 Test  #1: test_array .......................***Failed    0.01 sec
[..s/src/test_array.cpp:11 (main)] array1 = {0, 0, 0} (std::__1::array<int, 3>)
[..s/src/test_array.cpp:12 (main)] array2 = {1, 2, 3} (std::__1::array<int, 3>)
[..s/src/test_array.cpp:14 (main)] array3 = {0, 0, 0} (const std::__1::array<int, 3>)
[..s/src/test_array.cpp:15 (main)] array4 = {1, 2, 3} (const std::__1::array<int, 3>)
[..s/src/test_array.cpp:17 (main)] std::array<int, 3>{} = {0, 0, 0} (std::__1::array<int, 3>)
[..s/src/test_array.cpp:18 (main)] std::array<int, 3>{1, 2, 3} = {1, 2, 3} (std::__1::array<int, 3>)
Assertion 'lhs == rhs' failed!
[..cro/tests/src/test.h:7 (assert_eq)] lhs = "std::__1::array<int, 3>" (const std::string&)
[..cro/tests/src/test.h:8 (assert_eq)] std::array<int, 3>
```

## Questions
- Should I keep this failing test or not?
- Should I open a new issue?

**Comment by xorz57** [comment]
I also found out that AppVeyor build fails only if the last command fails. That being said, I moved the tests to the bottom.

**Comment by xorz57** [comment]
I finally fixed `appveyor.yml`! The build was succesful even though we had 1 test failing. Now everything works as intended. :fireworks:

**Comment by xorz57** [comment]
Resolved bug with `std::array`. All tests pass now! :fireworks:

**Comment by xorz57** [comment]
I also tried to build with VS2013 and VS2015 but there are some errors.
I thaught you midght want to take a look at [this](https://ci.appveyor.com/project/xorz57/dbg-macro/builds/25489870)

**Comment by xorz57** [comment]
I also added 2 ubuntu xenial images with clang installed.

**Comment by xorz57** [comment]
Resolved bug with `std::list`.
All tests **PASS**! :fireworks:

**Comment by xorz57** [comment]
@sharkdp I am done.

**Comment by sharkdp** [comment]
> I hope we can merge this one before merge conflicts start appearing again disappointed

Please consider splitting unrelated changes (windows compatibility, cmake updates, cmake install updates, split test cases, add new unit tests, add new examples, etc. etc.) across multiple PRs in the future. This PR is increasingly hard to review and I would have really liked to discuss some of the things you changed in more detail. But now everything is woven into one big changeset that is hard to split.

That being said, I really appreciate all the work you put into this PR!! I agree with most of the changes, but I may actually revert some of them after merging this PR. For example, I'm not sure if all of the new unit tests really add any value in terms of test coverage. Some of the tests that you removed were specifically designed to test very specific cases (like a `std::vector<…>` with *exactly* 5 elements).

If you find yourself being afraid of merge conflicts, try to actually perform a few manual merges and I think you will find that it's usually not bad at all. Still, splitting changes across multiple branches/PRs typically helps with both the review process and the potential conflicts.

In any case, Thank you very much!

**Comment by sharkdp** [comment]
> it would be nice if you pushed a `github release` for each stable version. That way we can use a `direct link` to the header without having to `clone` or `copy & paste` the header's source code

That's something I typically do for all of my projects, and certainly also a good idea for this one :+1:

File: include/dbg_macro/dbg.h [hunk 2]
```diff
@@ -29,23 +29,36 @@ License (MIT):
 #ifndef DBG_MACRO_DBG_H
 #define DBG_MACRO_DBG_H
 
-#pragma message("WARNING: the 'dbg.h' header is included in your code base")
```
Comment (sharkdp) [inline include/dbg_macro/dbg.h]: why was this removed?

File: include/dbg_macro/dbg.h [hunk 3]
```diff
@@ -199,10 +222,9 @@ struct has_begin_end_size {
 // Specializations of "pretty_print"
 
 template <typename T>
-typename std::enable_if<!has_begin_end_size<T>::value &&
-                            !std::is_enum<T>::value,
```
Comment (sharkdp) [inline include/dbg_macro/dbg.h]: why was the `!std::is_enum<T>::value` check removed? Is it not necessary?

Comment (sharkdp) [reply include/dbg_macro/dbg.h]: Oh, you actually just remove the test case :-(

File: include/dbg_macro/dbg.h [hunk 4]
```diff
@@ -229,7 +251,7 @@ bool pretty_print(std::ostream& stream, P* const& value) {
   return true;
 }
 
-template <size_t N>
```
Comment (sharkdp) [inline include/dbg_macro/dbg.h]: why was this changed back to size_t? Is that required for Windows compatibility?

**Comment by sharkdp** [comment]
I'm not going to merge this because I can't really tell anymore what kind of tests you removed/changed.

File: include/dbg_macro/dbg.h (hunk 2 above)
Comment (fgallegosalido) [reply include/dbg_macro/dbg.h]: I guess it's because it can get annoying getting that message in every compilation.

Comment (sharkdp) [reply include/dbg_macro/dbg.h]: #37

Comment (xorz57) [reply include/dbg_macro/dbg.h]: Yeah I have to add it again.

**Comment by sharkdp** [comment]
There are new merge conflicts here. I'm happy to help resolving them.

**Comment by xorz57** [comment]
@sharkdp Anyway! I will close this PR and I will probably open a new PR starting with CMake changes. Unfortunately, I have spent a lot of time for this PR so I need to take a break for some days. I agree that this PR is bloated and that I changed some of the test cases but some of it was done intentionaly (for example the tests for `std::array` and `std::list` were did not cover all cases). That being said, I think I have to open issues first and then make a **SMALL** PR for each issue.But first, I really want to have a clean cmake project structure.
